### DISCUSSION TITLE
Use `fastlane pilot builds` command from fastfile

### DISCUSSION THREAD

**Discussion by SumanDas94** [fastlane/fastlane#20618] [General]
I have a requirement to get the list of active threads from TF. 
The command `fastlane pilot builds` can help me achieve that but I am not sure how to use it as a fastfile action. 
Calling it using `sh` action would ask me for apple ID, password etc interactively and I need to run it on CI. 
Is there a way to use this functionality as an action in fastfile coupled with action `app_store_connect_api_key` action for API token or is there a way to get TF active thread list and their details via the fastfile ?

**Comment by SumanDas94** [comment]
@crazymanish can you help here ?

**Reply by crazymanish** [reply]
You should be able to use these couple of lines of code inside your Fastfile directly and this take/supports all Pilots input params that mean you should be able to use the app_store_connect_api_key also.
https://github.com/fastlane/fastlane/blob/ff41f3556f1baf8e700477e50f65be728780aa94/pilot/lib/pilot/commands_generator.rb#L87-L92

please let me know if this will help.

**Reply by SumanDas94** [reply]
@crazymanish `uninitialized constant Fastlane::FastFile::Pilot` getting this while trying to use the snippet you mentioned in the fastfile

**Reply by SumanDas94** [reply]
`fastlane_require 'Pilot'` i think this will solve that.

**Reply by crazymanish** [reply]
indeed! you can just use `require 'pilot'` top of your fastfile to use pilot directly

**Reply by SumanDas94** [reply]
@crazymanish this didn't work as I think `app_store_connect_api_key` doesn't work for pilot builds command. More on that in this thread https://github.com/fastlane/fastlane/issues/18650
Do you know how can we perform `fastlane pilot builds` command functionality
 on CI then . Using AppID + Password will prompt for two factor authentication i believe

**Reply by crazymanish** [reply]
people might have found/build a way to pass 2FA with CI.
i never needed that so never did that magic myself in our CI...but i think, below can help you
- https://github.com/fastlane/fastlane/blob/master/spaceship/docs/Authentication.md
and 
- https://github.com/fastlane/fastlane/discussions/17059

**Reply by SumanDas94** [reply]
hey @crazymanish i have opened this discussion thread https://github.com/fastlane/fastlane/discussions/20638. Do you know the right person to tag for this ?

**Reply by crazymanish** [reply]
i don't know the right person maybe below article can help

https://www.joshholtz.com/blog/2021/02/17/apples-2fa-with-fastlane.html
<img width="626" alt="Screenshot 2022-09-11 at 11 56 43" src="https://user-images.githubusercontent.com/5364500/189521625-4ee0e591-d84b-4d88-adfb-a0631e47f3ce.png">

**Reply by crazymanish** [reply]
other solution to enter the 2FA code 

https://drobinin.com/posts/keep-using-fastlane-with-apples-2fa-changes/#%F0%9F%92%AC-sms-parsing

**Reply by SumanDas94** [reply]
My query is actually coming from this section in the docs

```
Option 2: Migrate to App Store Connect API Key
This is my recommended approach if you can make this work for you as migrating from Apple ID to the App Store Connect API Key should be a minimal change for most users.

Instead of using the username option or FASTLANE_USER/FASTLANE_PASSWORD environment variables, you would use the api_key or api_key_path options or the APP_STORE_CONNECT_API_KEY or APP_STORE_CONNECT_API_KEY_PATH variables 😊

A majority of the tools/actions have support for these new options/environment variables:

Tools
cert
deliver
match
pilot
precheck
sigh
```

it says the pilot tool is able to use App Store Connect API Key but as i mentioned in the https://github.com/fastlane/fastlane/discussions/20638 `fastlane pilot builds` is not working with that and https://github.com/fastlane/fastlane/issues/18650 says it is not possible to use app store connect key altogether on this functionality

**Reply by SumanDas94** [reply]
Even the official docs of fastlane says the same https://docs.fastlane.tools/app-store-connect-api/

**Reply by crazymanish** [reply]
it seems **Apple** does not support all pilot features using the App Store Connect API Key. i.e fastlane pilot builds

**Reply by crazymanish** [reply]
and if Apple does support the pilot builds feature using App Store Connect API Key and Fastlane is not supporting that then please feel free to open PR! because that way we can help other people also who are looking for such features.

**Comment by shepting** [comment]
This exact use-case is something that I was trying to get working yesterday & today. Thanks for all of the investigation so far.
